crates/worker/src/partition_processor_manager/mod.rs

```diff
@@ -647,6 +656,52 @@ impl PartitionProcessorManager {
         }
     }
 
+    fn request_partition_snapshots(&mut self) {
+        let Some(records_per_snapshot) = self
+            .updateable_config
+            .live_load()
+            .worker
+            .snapshots
+            .snapshot_interval_num_records
+        else {
+            return;
+        };
+
+        for (partition_id, state) in self.processor_states.iter() {
+            let status = state.partition_processor_status();
+            match status {
+                Some(status)
+                    if status.replay_status == ReplayStatus::Active
+                        && status.last_applied_log_lsn.unwrap_or(Lsn::INVALID)
+                            >= status
+                                .last_archived_log_lsn
+                                .unwrap_or(Lsn::OLDEST)
+                                .add(Lsn::from(records_per_snapshot.get())) =>
```

Comment:
Should we also check that we are the leader?

Reply:
How will the snapshotting behave in the presence of leadership changes? It looks that a new leader would immediately create a new snapshot if the `last_applied_log_lsn` is larger than `recrods_per_snapshot`.

Reply:
We guard upstream whether this PP is a leader at the time that the request comes, so I don't bother to check again here - partly because we might change our mind in the future.

We don't currently synchronize the archived LSN across PPs, so yes, when leadership changes, the new leader will only consider _its_ acrhived LSN value and might decide to take a snapshot right away.

In a future PR (draft here: https://github.com/restatedev/restate/pull/2288/), I introduce the ability to retrieve the latest snapshot from the repository. When we have that in place, can update the PPs to refresh. This way, if you have an object store-based repo, they'll all converge. I haven't thought about it much more deeply than this though.

Reply:
"upstream" == "downstream" (in the partition processor itself)? So I guess you are going to change this in the PR where we are reworking where the snapshot is taken?

Reply:
I'm sorry, I'm being a dummy! When manual snapshots are requested from the Admin node, the cluster control service will only send requests to leader nodes. That's what I meant by "upstream" - but that's totally irrelevant for periodic snapshots!

To your original question, I've added back a condition that the current PP's effective mode has to be leader for a periodic snapshot to be requested to be created on it. This was here initially but mistakenly removed it, blame Friday afternoon brain!

With the removal of the FSM variable, we will temporarily be producing new snapshot after either a cold-start of the leader, or a leadership change from a follower to a leader, where the original snapshot is very outdated. I'll correct this later in the PR series.

Reply:
Maybe it is possible that once a PP becomes leader we take its current applied lsn as the basis for deciding on snapshots. That way a new leader needs to process another min_num_of_records_per_snapshot before a new snapshot gets created but we wouldn't create a snapshot that has already been created by the previous leader.